File: DailyModule/src/main/java/com/qihuan/dailymodule/view/DailyFragment.java
package com.qihuan.dailymodule.view;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.alibaba.android.arouter.facade.annotation.Route;
import com.alibaba.android.arouter.launcher.ARouter;
import com.chad.library.adapter.base.BaseQuickAdapter;
import com.qihuan.commonmodule.base.BaseFragment;
import com.qihuan.commonmodule.imageloader.ImageLoader;
import com.qihuan.commonmodule.imageloader.strategy.GlideStrategy;
import com.qihuan.commonmodule.router.Router;
import com.qihuan.dailymodule.R;
import com.qihuan.dailymodule.contract.DailyContract;
import com.qihuan.dailymodule.model.bean.DailyItemBean;
import com.qihuan.dailymodule.model.bean.TopStoryBean;
import com.qihuan.dailymodule.presenter.DailyPresenter;
import com.qihuan.dailymodule.view.adapter.DailyAdapter;

import java.util.List;

import cn.bingoogolapple.bgabanner.BGABanner;

/**
 * DailyFragment
 *
 * @author Qi
 */
@Route(path = Router.DAILY_FRAGMENT)
public class DailyFragment extends BaseFragment implements
    DailyContract.View,
    SwipeRefreshLayout.OnRefreshListener,
    BaseQuickAdapter.OnItemClickListener,
    BaseQuickAdapter.RequestLoadMoreListener,
    BGABanner.Adapter<RelativeLayout, TopStoryBean>,
    BGABanner.Delegate<RelativeLayout, TopStoryBean> {

    private SwipeRefreshLayout refreshLayout;
    private BGABanner bannerView;
    private DailyPresenter presenter;
    private DailyAdapter dailyAdapter;

    public static DailyFragment newInstance() {
        return new DailyFragment();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        presenter = new DailyPresenter();
        presenter.attachView(this);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.detachView();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_news, container, false);
        initView(view);
        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        bannerView.startAutoPlay();
    }

    @Override
    public void onPause() {
        super.onPause();
        bannerView.stopAutoPlay();
    }

    private void initView(View view) {
        RecyclerView rvList = view.findViewById(R.id.rv_list);
        refreshLayout = view.findViewById(R.id.refresh_layout);

        refreshLayout.setOnRefreshListener(this);
        dailyAdapter = new DailyAdapter();
        dailyAdapter.openLoadAnimation(BaseQuickAdapter.SLIDEIN_BOTTOM);
        dailyAdapter.setOnItemClickListener(this);
        dailyAdapter.setOnLoadMoreListener(this, rvList);
        rvList.setLayoutManager(new LinearLayoutManager(getContext()));
        rvList.setAdapter(dailyAdapter);

        bannerView = (BGABanner) LayoutInflater.from(getContext()).inflate(R.layout.layout_banner, rvList, false);
        bannerView.setAdapter(this);
        bannerView.setDelegate(this);
        dailyAdapter.addHeaderView(bannerView);

        presenter.getLatestDaily();
    }

    @Override
    public void onRefresh() {
        presenter.getLatestDaily();
    }

    @Override
    public void onLoadMoreRequested() {
        presenter.getBeforeDaily();
    }

    @Override
    public void onItemClick(BaseQuickAdapter adapter, View view, int position) {
        DailyItemBean dailyItemBean = (DailyItemBean) adapter.getData().get(position);
        if (dailyItemBean.isHeader) {
            return;
        }
        if (dailyItemBean.t == null) {
            return;
        }
        ARouter.getInstance()
            .build(Router.DAILY_DET_ACTIVITY)
            .withInt("id", dailyItemBean.t.getId())
            .navigation();
    }

    @Override
    public void fillBannerItem(BGABanner banner, RelativeLayout itemView, TopStoryBean model, int position) {
        ImageView ivBanner = itemView.findViewById(R.id.iv_banner);
        TextView tvBanner = itemView.findViewById(R.id.tv_banner);

        ImageLoader.INSTANCE
            .strategy(new GlideStrategy())
            .with(getContext())
            .load(model.getImage())
            .into(ivBanner);
        tvBanner.setText(model.getTitle());
    }

    @Override
    public void onBannerItemClick(BGABanner banner, RelativeLayout itemView, TopStoryBean model, int position) {
        ARouter.getInstance()
            .build(Router.DAILY_DET_ACTIVITY)
            .withInt("id", model.getId())
            .navigation();
    }

    @Override
    public void latestDaily(List<TopStoryBean> topList) {
        bannerView.setData(R.layout.item_daily_banner, topList, null);
    }

    @Override
    public void beforeDaily(boolean isRefresh, List<DailyItemBean> dailyList) {
        if (isRefresh) {
            dailyAdapter.setNewData(dailyList);
        } else {
            dailyAdapter.addData(dailyList);
        }
    }

    @Override
    public void onRefreshEnd() {
        refreshLayout.setRefreshing(false);
    }

    @Override
    public void onLoadMoreEnd(boolean success) {
        if (success) {
            dailyAdapter.loadMoreComplete();
        } else {
            dailyAdapter.loadMoreFail();
        }
    }

    @Override
    public void showLoading() {

    }

    @Override
    public void hideLoading() {

    }

    @Override
    public void showError(String errorMsg) {

    }
}

File: module_daily/src/main/java/com/qihuan/dailymodule/view/DailyFragment.kt
package com.qihuan.dailymodule.view

import android.os.Bundle
import androidx.recyclerview.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.alibaba.android.arouter.facade.annotation.Route
import com.qihuan.commonmodule.base.BaseMvpFragment
import com.qihuan.commonmodule.bus.BindEventBus
import com.qihuan.commonmodule.bus.event.RefreshEvent
import com.qihuan.commonmodule.router.Routes
import com.qihuan.dailymodule.R
import com.qihuan.dailymodule.contract.DailyContract
import com.qihuan.dailymodule.model.bean.StoryBean
import com.qihuan.dailymodule.model.bean.TopStoryBean
import com.qihuan.dailymodule.presenter.DailyPresenter
import com.qihuan.dailymodule.view.cell.BannerCell
import com.qihuan.dailymodule.view.cell.DailyCell
import com.qihuan.dailymodule.view.cell.DailySectionCell
import kotlinx.android.synthetic.main.fragment_news.*
import me.drakeet.multitype.Items
import me.drakeet.multitype.MultiTypeAdapter
import me.drakeet.multitype.register
import org.greenrobot.eventbus.Subscribe

/**
 * DailyFragment
 *
 * @author Qi
 */
@BindEventBus
@Route(path = Routes.DAILY_FRAGMENT)
class DailyFragment : BaseMvpFragment<DailyContract.View, DailyContract.Presenter>(), DailyContract.View {

    private var dailyAdapter: MultiTypeAdapter? = null
    private var linearLayoutManager: androidx.recyclerview.widget.LinearLayoutManager? = null
    private val itemList: Items = Items()

    override fun initPresenter(): DailyContract.Presenter {
        return DailyPresenter()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_news, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // list
        linearLayoutManager = androidx.recyclerview.widget.LinearLayoutManager(context)
        rv_list.layoutManager = linearLayoutManager

        // adapter
        dailyAdapter = MultiTypeAdapter(itemList)
        dailyAdapter?.register(Array<TopStoryBean>::class, BannerCell())
        dailyAdapter?.register(StoryBean::class, DailyCell())
        dailyAdapter?.register(String::class, DailySectionCell())
        rv_list.adapter = dailyAdapter

        // refresh layout
        refresh_layout.apply {
            setOnRefreshListener { mPresenter.getLatestDaily() }
            setOnLoadMoreListener { mPresenter.getBeforeDaily() }
            autoRefresh()
        }
    }

    override fun onBannerData(itemList: Array<TopStoryBean>) {
        this.itemList.add(itemList)
        dailyAdapter?.notifyItemInserted(0)
    }

    override fun onDailyData(itemList: Array<StoryBean>) {
        val start = this.itemList.size
        this.itemList.addAll(itemList)
        dailyAdapter?.notifyItemRangeInserted(start, itemList.size)
    }

    override fun onDailySectionData(item: String) {
        val start = this.itemList.size
        this.itemList.add(item)
        dailyAdapter?.notifyItemInserted(start)
    }

    override fun onRefreshStart() {
        val count = this.itemList.size
        this.itemList.clear()
        dailyAdapter?.notifyItemRangeRemoved(0, count)
    }

    override fun onRefreshEnd(success: Boolean) {
        refresh_layout.finishRefresh(success)
    }

    override fun onLoadMoreEnd(success: Boolean) {
        refresh_layout.finishLoadMore(success)
    }

    @Subscribe
    fun onRefreshEvent(@Suppress("UNUSED_PARAMETER") refreshEvent: RefreshEvent) {
        linearLayoutManager?.run {
            val visibleItemPosition = findFirstCompletelyVisibleItemPosition()
            if (visibleItemPosition == 0) {
                // 刷新
                refresh_layout.autoRefresh()
                return
            }
            // 超过 20 条, 先滚动到20条, 再平滑滚动
            if (visibleItemPosition > 20) {
                rv_list.scrollToPosition(20)
            }
            rv_list.smoothScrollToPosition(0)
        }
    }

    override fun showLoading() {

    }

    override fun hideLoading() {

    }

    override fun showError(errorMsg: String) {

    }

}
